### ISSUE TITLE
Should #{#entityName} pick the domain type for the repository or the return type for a method? [DATAJPA-1046]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-jpa#1390]
**[Manish](https://jira.spring.io/secure/ViewProfile.jspa?name=manish.in.java)** opened **[DATAJPA-1046](https://jira.spring.io/browse/DATAJPA-1046?redirect=false)** and commented

Consider the following classes and repository interfaces:

```
    @Entity
    class Garage {}

    @Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
    @MappedSuperclass
    abstract class Vehicle {
      @ManyToOne private Garage garage;
    }

    @Entity
    class Motorcycle extends Vehicle {}

    @Entity
    class Truck extends Vehicle {}

    @NoRepositoryBean
    interface VehicleRepository<T extends Vehicle> extends JpaRepository<T, Long> {
      @Query("SELECT DISTINCT v.garage FROM #{#entityName} v")
      List<Garage> findAllDistinctGarages();
    }

    interface MotorcycleRepository extends VehicleRepository<Motorcycle> {}

    interface TruckRepository extends VehicleRepository<Truck>
```

I would expect this code to work based on [the documentation](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#jpa.query.spel-expressions), assuming that `entityName` picks `Truck` or `Motorcycle` at runtime. However, this fails with the error:

```
org.hibernate.QueryException: could not resolve property: garage of: Garage [SELECT DISTINCT v.garage FROM Garage v]
```

It seems that `entityName` is being resolved to the return type of the repository method `findAllDistinctGarages` instead of the domain type for the repository, which is how it is explained in the documentation.

What should we expect `entityName` to resolve to? If it should indeed resolve to the return type of the repository method, is there a way to get this scenario to work, given that the method cannot be taken to `GarageRepository` due to multiple `Vehicle` tables involved?

A link to a full sample demonstrating the problem is attached with this issue

---

**Affects:** 1.10.6 (Hopper SR6)

**Reference URL:** https://github.com/manish-in-java/stackoverflow-questions/tree/master/41701838

**Comment by gregturn** [comment]
A) The original issue regarding #{#entityName} and an off-repository type is not an issue. I verified it and the OP did too.

B) The secondary issue is also not an issue, as I added the same operation as the OP did to `ModelRepository`, and it worked great.

NOTE: Sometimes, we write test cases where ALL the types are buried inside the test class. This seems to break this entire test setup. However, with all the entity types and repositories, everything worked for me.
